# Домашнее задание 
# Тема: Действия с мультииндексами в датафреймах

In [23]:
import pandas as pd
import numpy as np

## Задание 1
### Постройте датафрейм заполненный нулями так, как это показано в ячейке ниже. Первые две колонки являются мультииндексом. Для заполнения датафрейма нулями воспользуйтесь функцией zeros из библиотеки numpy

In [24]:
df1_result = pd.DataFrame(index=[['Группа 1', 'Группа 1', 'Группа 2', 'Группа 2'],['Альфа', 'Бета', 'Альфа', 'Бета']],
                          columns=[['Владивосток', 'Владивосток', 'Казань', 'Новосибирск', 'Новосибирск'],['Кол-во', 'Сумма', 'Сумма', 'Кол-во', 'Сумма']], data=np.zeros((4, 5)))
print(df1_result)

               Владивосток       Казань Новосибирск      
                    Кол-во Сумма  Сумма      Кол-во Сумма
Группа 1 Альфа         0.0   0.0    0.0         0.0   0.0
         Бета          0.0   0.0    0.0         0.0   0.0
Группа 2 Альфа         0.0   0.0    0.0         0.0   0.0
         Бета          0.0   0.0    0.0         0.0   0.0


### Получившийся датафрейм сохраните в csv файл и загрузите на степик

In [25]:
df1_result.to_csv('df1_result.csv')

## Задание 2
### Загрузите датафрейм из файла task5.csv. Поменяйте местами уровни так чтобы сначала шел тип парковки (платная/бесплатная), затем округ и район. Отсортируйте так чтобы сначала шли бесплатные парковки в алфавитном порядке по округу, а внутри округа в алфавитном порядке по району*. 

#### *Некоторые названия районов записаны наоборот: вместо район Нижегородский записано Нижегородский район. Из-за этого сортировка по алфавиту будет идти не совсем так, как вы ожидали. Оставим результат сортировки по алфавиту как есть (переносить слово район для корректной сортировки не будем).

In [26]:
df2_task = pd.read_csv('/Users/magewade/Desktop/Python/Pandas/Homework_multiindex_2/task5.csv', index_col=[0, 1, 2], header=[0, 1]) # не редактируйте ячейку
df2_task = df2_task.swaplevel(2,0)
df2_task = df2_task.swaplevel(1,2)
df2_task = df2_task.sort_index(level=[0, 1, 2])

In [27]:
df2_task

Capacity   
                                                                            count   
Бесплатная Восточный административный округ Нижегородский район                 1  \
                                            район Богородское                   2   
                                            район Вешняки                       5   
                                            район Восточное Измайлово           1   
                                            район Гольяново                     2   
...                                                                           ...   
Платная    Южный административный округ     район Орехово-Борисово Южное       10   
                                            район Царицыно                      4   
                                            район Чертаново Северное            2   
                                            район Чертаново Центральное         3   
                                            район Чертаново Южное               4   

                                                                               
                                                                         sum   
Бесплатная Восточный административный округ Нижегородский район            6  \
                                            район Богородское             14   
                                            район Вешняки                 35   
                                            район Восточное Измайлово      8   
                                            район Гольяново                9   
...                                                                       ..   
Платная    Южный административный округ     район Орехово-Борисово Южное  85   
                                            район Царицыно                38   
                                            район Чертаново Северное      20   
                                            район Чертаново Центральное   21   
                                            район Чертаново Южное         36   

                                                                         rating   
                                                                            min   
Бесплатная Восточный административный округ Нижегородский район               8  \
                                            район Богородское                 8   
                                            район Вешняки                     2   
                                            район Восточное Измайлово         6   
                                            район Гольяново                   0   
...                                                                         ...   
Платная    Южный административный округ     район Орехово-Борисово Южное      0   
                                            район Царицыно                    2   
                                            район Чертаново Северное          2   
                                            район Чертаново Центральное       3   
                                            район Чертаново Южное             0   

                                                                               
                                                                         max   
Бесплатная Восточный административный округ Нижегородский район            8  \
                                            район Богородское             10   
                                            район Вешняки                 10   
                                            район Восточное Измайлово      6   
                                            район Гольяново               10   
...                                                                       ..   
Платная    Южный административный округ     район Орехово-Борисово Южное   7   
                                            район Царицыно                10   
                                            ра

#### Обратите внимание, наш датафрейм немного усложнился: у мультииндекса строк пропали названия для уровней (добавлять их не стоит) и появился мультииндекс для колонок. Теперь в датафрейме содержится дополнительная информация о рейтинге парковки: маинимальное, максимальное и среднее значение

In [28]:
df2_result = df2_task[:27]

### Сохраните на степик только первые 27 записей датафрейма.

In [37]:
df2_result.to_csv('df2_result.csv')

## Задание 3
### Загрузите датафрейм из файла task5.csv. Из мультииндекса строк сбросьте уровень в котором содержится информация о типе парковки. Сохраните данные из этого уровня в отдельную колонку под названием `type`. Из целого датафрейма вырезаем только данные о рейтинге, сортируем записи по среднему рейтингу в порядке убывания и загружаем на степик ТОП-10 записей.

In [29]:
df3_task = pd.read_csv('/Users/magewade/Desktop/Python/Pandas/Homework_multiindex_2/task5.csv', index_col=[0, 1, 2], header=[0, 1]) # не редактируйте ячейку
df3_task

Capacity   
                                                                           count   
Восточный административный округ Нижегородский район         Бесплатная        1  \
                                 район Богородское           Бесплатная        2   
                                                             Платная           1   
                                 район Вешняки               Бесплатная        5   
                                                             Платная           3   
...                                                                          ...   
Южный административный округ     район Чертаново Северное    Платная           2   
                                 район Чертаново Центральное Бесплатная        7   
                                                             Платная           3   
                                 район Чертаново Южное       Бесплатная        9   
                                                             Платная           4   

                                                                              
                                                                        sum   
Восточный административный округ Нижегородский район         Бесплатная   6  \
                                 район Богородское           Бесплатная  14   
                                                             Платная      8   
                                 район Вешняки               Бесплатная  35   
                                                             Платная     17   
...                                                                      ..   
Южный административный округ     район Чертаново Северное    Платная     20   
                                 район Чертаново Центральное Бесплатная  68   
                                                             Платная     21   
                                 район Чертаново Южное       Бесплатная  86   
                                                             Платная     36   

                                                                        rating   
                                                                           min   
Восточный административный округ Нижегородский район         Бесплатная      8  \
                                 район Богородское           Бесплатная      8   
                                                             Платная         0   
                                 район Вешняки               Бесплатная      2   
                                                             Платная         0   
...                                                                        ...   
Южный административный округ     район Чертаново Северное    Платная         2   
                                 район Чертаново Центральное Бесплатная      0   
                                                             Платная         3   
                                 район Чертаново Южное       Бесплатная      3   
                                                             Платная         0   

                                                                              
                                                                        max   
Восточный административный округ Нижегородский район         Бесплатная   8  \
                                 район Богородское           Бесплатная  10   
                                                             Платная      0   
                                 район Вешняки               Бесплатная  10   
                                                             Платная      9   
...                                                                      ..   
Южный административный округ     район Чертаново Северное    Платная      6   
                                 район Чертаново Центральное Бесплатная   9   
                                                             Платная      9   
               

In [31]:
# print(df3_task[df3_task['rating']['min'] >= 5])
df3_task_sum = df3_task.sum(axis=1)
df3_task_sum

Восточный административный округ  Нижегородский район          Бесплатная     31.000000
                                  район Богородское            Бесплатная     43.000000
                                                               Платная         9.000000
                                  район Вешняки                Бесплатная     58.800000
                                                               Платная        33.666667
                                                                                ...    
Южный административный округ      район Чертаново Северное     Платная        34.000000
                                  район Чертаново Центральное  Бесплатная     87.571429
                                                               Платная        42.000000
                                  район Чертаново Южное        Бесплатная    114.666667
                                                               Платная        54.000000
Length: 247, dtype: float64

In [32]:
df2_result = df3_task.reset_index(level=2)
df2_result = df2_result.drop(columns=("Capacity", "count"), axis=1)
df2_result = df2_result.drop(columns=("Capacity", "sum"), axis=1)
df2_result.columns = df2_result.columns.droplevel(level=0)
df2_result.columns.values[0] = 'type'
df2_result = df2_result.sort_values(by="mean", ascending=False)

In [33]:
df2_result = df2_result[:10]

In [34]:
df2_result

type  min   
Юго-Восточный административный округ  район Некрасовка      Бесплатная   10  \
Зеленоградский административный округ район Силино             Платная   10   
Северный административный округ       район Сокол              Платная   10   
Западный административный округ       район Марьино            Платная   10   
                                      район Филёвский Парк  Бесплатная   10   
Юго-Восточный административный округ  район Печатники       Бесплатная    9   
Западный административный округ       Можайский район       Бесплатная    9   
                                      район Дорогомилово    Бесплатная    9   
Юго-Восточный административный округ  Гагаринский район     Бесплатная    9   
Новомосковский административный округ поселение Сосенское      Платная    9   

                                                            max  mean  
Юго-Восточный административный округ  район Некрасовка       10  10.0  
Зеленоградский административный округ район Силино           10  10.0  
Северный административный округ       район Сокол            10  10.0  
Западный административный округ       район Марьино          10  10.0  
                                      район Филёвский Парк   10  10.0  
Юго-Восточный административный округ  район Печатники         9   9.0  
Западный административный округ       Можайский район         9   9.0  
                                      район Дорогомилово      9   9.0  
Юго-Восточный административный округ  Гагаринский район       9   9.0  
Новомосковский административный округ поселение Сосенское     9   9.0

### Сохраните и отправьте файл на степик.

In [64]:
df2_result.to_csv('df3_result.csv')

## Задание 4 
### В одном из прошлых заданий мы выбирали все районы восточного административного округа и искали где больше всего мест на бесплатных парковках. Загрузите датафрейм из файла task5.csv и попробуйте найти в каком округе больше всего парковок (и платных и бесплатных вместе). Учитывайте только те районы, где средний рейтинг парковок больше или равен 5 

In [35]:
df4_task = pd.read_csv('/Users/magewade/Desktop/Python/Pandas/Homework_multiindex_2/task5.csv', index_col=[0, 1, 2], header=[0, 1]) # не редактируйте ячейку

In [36]:
df4_task

Capacity   
                                                                           count   
Восточный административный округ Нижегородский район         Бесплатная        1  \
                                 район Богородское           Бесплатная        2   
                                                             Платная           1   
                                 район Вешняки               Бесплатная        5   
                                                             Платная           3   
...                                                                          ...   
Южный административный округ     район Чертаново Северное    Платная           2   
                                 район Чертаново Центральное Бесплатная        7   
                                                             Платная           3   
                                 район Чертаново Южное       Бесплатная        9   
                                                             Платная           4   

                                                                              
                                                                        sum   
Восточный административный округ Нижегородский район         Бесплатная   6  \
                                 район Богородское           Бесплатная  14   
                                                             Платная      8   
                                 район Вешняки               Бесплатная  35   
                                                             Платная     17   
...                                                                      ..   
Южный административный округ     район Чертаново Северное    Платная     20   
                                 район Чертаново Центральное Бесплатная  68   
                                                             Платная     21   
                                 район Чертаново Южное       Бесплатная  86   
                                                             Платная     36   

                                                                        rating   
                                                                           min   
Восточный административный округ Нижегородский район         Бесплатная      8  \
                                 район Богородское           Бесплатная      8   
                                                             Платная         0   
                                 район Вешняки               Бесплатная      2   
                                                             Платная         0   
...                                                                        ...   
Южный административный округ     район Чертаново Северное    Платная         2   
                                 район Чертаново Центральное Бесплатная      0   
                                                             Платная         3   
                                 район Чертаново Южное       Бесплатная      3   
                                                             Платная         0   

                                                                              
                                                                        max   
Восточный административный округ Нижегородский район         Бесплатная   8  \
                                 район Богородское           Бесплатная  10   
                                                             Платная      0   
                                 район Вешняки               Бесплатная  10   
                                                             Платная      9   
...                                                                      ..   
Южный административный округ     район Чертаново Северное    Платная      6   
                                 район Чертаново Центральное Бесплатная   9   
                                                             Платная      9   
               

In [37]:
df = df4_task[df4_task["rating"]["mean"] >= 5]

In [38]:
result = df.groupby(level=0).sum()

In [40]:
result.sort_values(by=('Capacity', 'count'), ascending=False)

Capacity      rating                 
                                           count  sum    min  max        mean
Северо-Восточный административный округ      110  953     52  183  124.219170
Центральный административный округ            98  305     20  100   63.394824
Южный административный округ                  60  498     70  160  116.800000
Восточный административный округ              35  216     61  119   92.450000
Юго-Восточный административный округ          35  248     66  115   91.000000
Северный административный округ               32  255     81  124  104.150000
Западный административный округ               28  219     82  109   95.500000
Юго-Западный административный округ           27  241     26   66   49.150000
Северо-Западный административный округ        14   99     18   40   28.866667
Зеленоградский административный округ         13   94     19   36   28.100000
Новомосковский административный округ          8   96     20   35   27.500000

In [41]:
df

Capacity   
                                                                           count   
Восточный административный округ Нижегородский район         Бесплатная        1  \
                                 район Богородское           Бесплатная        2   
                                 район Вешняки               Бесплатная        5   
                                 район Восточное Измайлово   Бесплатная        1   
                                                             Платная           1   
...                                                                          ...   
Южный административный округ     район Царицыно              Платная           4   
                                 район Чертаново Северное    Бесплатная        6   
                                 район Чертаново Центральное Платная           3   
                                 район Чертаново Южное       Бесплатная        9   
                                                             Платная           4   

                                                                              
                                                                        sum   
Восточный административный округ Нижегородский район         Бесплатная   6  \
                                 район Богородское           Бесплатная  14   
                                 район Вешняки               Бесплатная  35   
                                 район Восточное Измайлово   Бесплатная   8   
                                                             Платная     10   
...                                                                      ..   
Южный административный округ     район Царицыно              Платная     38   
                                 район Чертаново Северное    Бесплатная  58   
                                 район Чертаново Центральное Платная     21   
                                 район Чертаново Южное       Бесплатная  86   
                                                             Платная     36   

                                                                        rating   
                                                                           min   
Восточный административный округ Нижегородский район         Бесплатная      8  \
                                 район Богородское           Бесплатная      8   
                                 район Вешняки               Бесплатная      2   
                                 район Восточное Измайлово   Бесплатная      6   
                                                             Платная         7   
...                                                                        ...   
Южный административный округ     район Царицыно              Платная         2   
                                 район Чертаново Северное    Бесплатная      2   
                                 район Чертаново Центральное Платная         3   
                                 район Чертаново Южное       Бесплатная      3   
                                                             Платная         0   

                                                                              
                                                                        max   
Восточный административный округ Нижегородский район         Бесплатная   8  \
                                 район Богородское           Бесплатная  10   
                                 район Вешняки               Бесплатная  10   
                                 район Восточное Измайлово   Бесплатная   6   
                                                             Платная      7   
...                                                                      ..   
Южный административный округ     район Царицыно              Платная     10   
                                 район Чертаново Северное    Бесплатная  10   
                                 район Чертаново Центральное Платная      9   
               

### Правильный ответ выберите на степике

## Задание 5
### Загрузите датафрейм из файла task5.csv. В этой задаче вы попробуете перейти от мультииндекса к обычному индексу с сохранением информации об уровнях. Сейчас на нулевом уровне колонки разбиваются на Capacity и rating, а на первом уровне Capacity и rating детализируются на count/sum и min/max/mean соответственно.
### Попытайтесь уйти от иерархии с сохранением информации о ней: сделайте обычный индекс с названием колонок:
#### Capacity_count, Capacity_sum, rating_min, rating_max, rating_mean

#### Не пытайтесь руками составить список и присвоить его т.к. на практике у вас может быть много колонок и руками перебирать будет долго. А еще можно что-то перепутать и тогда будет ошибка в последующих расчётах. Попробуйте перебрать значения мультииндекса в цикле и собрать тот самый список, который уже и присвоите обратно в датафрейм

In [42]:
df5_task = pd.read_csv('/Users/magewade/Desktop/Python/Pandas/Homework_multiindex_2/task5.csv', index_col=[0, 1, 2], header=[0, 1]) # не редактируйте ячейку

In [44]:
df5_task

Capacity   
                                                                           count   
Восточный административный округ Нижегородский район         Бесплатная        1  \
                                 район Богородское           Бесплатная        2   
                                                             Платная           1   
                                 район Вешняки               Бесплатная        5   
                                                             Платная           3   
...                                                                          ...   
Южный административный округ     район Чертаново Северное    Платная           2   
                                 район Чертаново Центральное Бесплатная        7   
                                                             Платная           3   
                                 район Чертаново Южное       Бесплатная        9   
                                                             Платная           4   

                                                                              
                                                                        sum   
Восточный административный округ Нижегородский район         Бесплатная   6  \
                                 район Богородское           Бесплатная  14   
                                                             Платная      8   
                                 район Вешняки               Бесплатная  35   
                                                             Платная     17   
...                                                                      ..   
Южный административный округ     район Чертаново Северное    Платная     20   
                                 район Чертаново Центральное Бесплатная  68   
                                                             Платная     21   
                                 район Чертаново Южное       Бесплатная  86   
                                                             Платная     36   

                                                                        rating   
                                                                           min   
Восточный административный округ Нижегородский район         Бесплатная      8  \
                                 район Богородское           Бесплатная      8   
                                                             Платная         0   
                                 район Вешняки               Бесплатная      2   
                                                             Платная         0   
...                                                                        ...   
Южный административный округ     район Чертаново Северное    Платная         2   
                                 район Чертаново Центральное Бесплатная      0   
                                                             Платная         3   
                                 район Чертаново Южное       Бесплатная      3   
                                                             Платная         0   

                                                                              
                                                                        max   
Восточный административный округ Нижегородский район         Бесплатная   8  \
                                 район Богородское           Бесплатная  10   
                                                             Платная      0   
                                 район Вешняки               Бесплатная  10   
                                                             Платная      9   
...                                                                      ..   
Южный административный округ     район Чертаново Северное    Платная      6   
                                 район Чертаново Центральное Бесплатная   9   
                                                             Платная      9   
               

In [45]:
import pandas as pd
# df5_result = df5_task.reset_index()
df5_result = df5_task.T.reset_index().T

df5_result = df5_result.drop(df5_result.index[0])
df5_result.columns = df5_result.iloc[0]
df5_result = df5_result.drop(df5_result.index[0])
df5_result.columns = ['Capacity_count', 'Capacity_sum', 'rating_min', 'rating_max', 'rating_mean']
df5_result[['Capacity_count', 'Capacity_sum', 'rating_min', 'rating_max']] = df5_result[['Capacity_count', 'Capacity_sum', 'rating_min', 'rating_max']].astype(int)

In [47]:
df5_result

Capacity_count   
Восточный административный округ Нижегородский район         Бесплатная               1  \
                                 район Богородское           Бесплатная               2   
                                                             Платная                  1   
                                 район Вешняки               Бесплатная               5   
                                                             Платная                  3   
...                                                                                 ...   
Южный административный округ     район Чертаново Северное    Платная                  2   
                                 район Чертаново Центральное Бесплатная               7   
                                                             Платная                  3   
                                 район Чертаново Южное       Бесплатная               9   
                                                             Платная                  4   

                                                                         Capacity_sum   
Восточный административный округ Нижегородский район         Бесплатная             6  \
                                 район Богородское           Бесплатная            14   
                                                             Платная                8   
                                 район Вешняки               Бесплатная            35   
                                                             Платная               17   
...                                                                               ...   
Южный административный округ     район Чертаново Северное    Платная               20   
                                 район Чертаново Центральное Бесплатная            68   
                                                             Платная               21   
                                 район Чертаново Южное       Бесплатная            86   
                                                             Платная               36   

                                                                         rating_min   
Восточный административный округ Нижегородский район         Бесплатная           8  \
                                 район Богородское           Бесплатная           8   
                                                             Платная              0   
                                 район Вешняки               Бесплатная           2   
                                                             Платная              0   
...                                                                             ...   
Южный административный округ     район Чертаново Северное    Платная              2   
                                 район Чертаново Центральное Бесплатная           0   
                                                             Платная              3   
                                 район Чертаново Южное       Бесплатная           3   
                                                             Платная              0   

                                                                         rating_max   
Восточный административный округ Нижегородский район         Бесплатная           8  \
                                 район Богородское           Бесплатная          10   
                                                             Платная              0   
                                 район Вешняки               Бесплатная          10   
                                                             Платная              9   
...                                                                             ...   
Южный административный округ     район Чертаново Северное    Платная              6   
                                 район Чертаново Центральное Бесплатная           9   
                                                             Платная              9

In [18]:

df5_result.to_csv("df5_task.csv")

### Сохраните и отправьте файл на степик.

In [124]:
df5_result.to_csv('df5_result.csv')

## Задача 6
### Загрузите данные из task5.csv. По всей Москве рассчитайте:
### -Общее количество бесплатных парковок,
### -Общее количество платных парковок,
### -Общее количество мест на бесплатных парковках,
### -Общее количество мест на платных парковках

### Введите в поле с ответом получившиеся значения через пробел. В ответе должны быть целые числа
### Пример: 10 12 90 110

In [48]:
df6_task = pd.read_csv('/Users/magewade/Desktop/Python/Pandas/Homework_multiindex_2/task5.csv', index_col=[0, 1, 2], header=[0, 1])

In [49]:
df6_task

Capacity   
                                                                           count   
Восточный административный округ Нижегородский район         Бесплатная        1  \
                                 район Богородское           Бесплатная        2   
                                                             Платная           1   
                                 район Вешняки               Бесплатная        5   
                                                             Платная           3   
...                                                                          ...   
Южный административный округ     район Чертаново Северное    Платная           2   
                                 район Чертаново Центральное Бесплатная        7   
                                                             Платная           3   
                                 район Чертаново Южное       Бесплатная        9   
                                                             Платная           4   

                                                                              
                                                                        sum   
Восточный административный округ Нижегородский район         Бесплатная   6  \
                                 район Богородское           Бесплатная  14   
                                                             Платная      8   
                                 район Вешняки               Бесплатная  35   
                                                             Платная     17   
...                                                                      ..   
Южный административный округ     район Чертаново Северное    Платная     20   
                                 район Чертаново Центральное Бесплатная  68   
                                                             Платная     21   
                                 район Чертаново Южное       Бесплатная  86   
                                                             Платная     36   

                                                                        rating   
                                                                           min   
Восточный административный округ Нижегородский район         Бесплатная      8  \
                                 район Богородское           Бесплатная      8   
                                                             Платная         0   
                                 район Вешняки               Бесплатная      2   
                                                             Платная         0   
...                                                                        ...   
Южный административный округ     район Чертаново Северное    Платная         2   
                                 район Чертаново Центральное Бесплатная      0   
                                                             Платная         3   
                                 район Чертаново Южное       Бесплатная      3   
                                                             Платная         0   

                                                                              
                                                                        max   
Восточный административный округ Нижегородский район         Бесплатная   8  \
                                 район Богородское           Бесплатная  10   
                                                             Платная      0   
                                 район Вешняки               Бесплатная  10   
                                                             Платная      9   
...                                                                      ..   
Южный административный округ     район Чертаново Северное    Платная      6   
                                 район Чертаново Центральное Бесплатная   9   
                                                             Платная      9   
               

In [4]:
import dtale

In [50]:
q_1 = df6_task.loc[:,:,'Бесплатная']
q_1 = q_1.sum()
q_1

Capacity  count     518.000000
          sum      3455.000000
rating    min       316.000000
          max       917.000000
          mean      610.116818
dtype: float64

In [51]:
q_2 = df6_task.loc[:,:,'Платная']
q_2 = q_2.sum()
q_2

Capacity  count     529.000000
          sum      3642.000000
rating    min       309.000000
          max       904.000000
          mean      607.048249
dtype: float64